### PR TITLE
Usability test results

### PR DESCRIPTION
**Pull request by tobi-v**
Implementation of usability test results

### CODE REVIEW

File: pilz_teleoperation/README.md [hunk 1]
```diff
@@ -20,20 +20,14 @@ Use following Syntax:
     ```
     MovementBindings:
       'w': <Twist> (e.g. {"linear": {"x": 1.0}})
-      '<CURSES_KEY_NAME>': <Twist> (e.g. KEY_LEFT)
     SettingBindings:
```
Comment (DamKoVosh) [inline pilz_teleoperation/README.md:25]: JointJogBindings missing.
Id suggest adding a hint for them as well

File: pilz_teleoperation/README.md [hunk 2]
```diff
@@ -20,20 +20,14 @@ Use following Syntax:
...
-      '+': <SettingName> (e.g. USE_XZ_PLANE)
+      '+': <SettingName> (e.g. INCREASE_LINEAR_VELOCITY)
     Description: |-
       Displayed Text e.g. to show current bindings
       With the '|-' the text can contain multiple rows
     ```
     
-    See [SetTeleopSettings.srv](./srv/SetTeleopSettings.srv) for available settings.
```
Comment (DamKoVosh) [inline pilz_teleoperation/README.md:31]: why is this removed?

File: pilz_teleoperation/README.md [hunk 3]
```diff
@@ -20,20 +20,14 @@ Use following Syntax:
...
-    
-    See [Curses key names](https://www.gnu.org/software/guile-ncurses/manual/html_node/Getting-characters-from-the-keyboard.html)
-    for name of special characters in curses.
-    
-    See [Default Keyboard Configuration](./config/keyboard_binding.yaml) for reference.
```
Comment (DamKoVosh) [inline pilz_teleoperation/README.md:36]: Same for default keyboard condiguration, id leave it in here.

File: pilz_teleoperation/config/keyboard_binding.yaml [hunk 4]
```diff
@@ -59,4 +60,4 @@ Description: |-
          up |    down : haul
   
   numpad : move inside the selected plane
-  , : change plane
+  , : change plane
```
Comment (DamKoVosh) [inline pilz_teleoperation/config/keyboard_binding.yaml]: why removing the newline at the eof?

File: pilz_teleoperation/launch/key_teleop.launch [hunk 5]
```diff
@@ -1,3 +1,19 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+You should have received a copy of the GNU Lesser General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+-->
+
 <launch>
     <node name="key_input" pkg="pilz_teleoperation" type="key_teleop.py" output="screen" />
 </launch>
```
Comment (DamKoVosh) [inline pilz_teleoperation/launch/key_teleop.launch]:
```suggestion
</launch>

```

File: pilz_teleoperation/src/pilz_teleoperation/__init__.py [hunk 6]
```diff
@@ -17,4 +17,4 @@
 from .teleoperation_input import TeleoperationInput
 from .teleoperation_window import TeleoperationWindow
 from .curses_key_input import CursesKeyInput
-from .terminal_text_window import TerminalTextWindow
+from .terminal_text_window import TerminalTextWindow
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/__init__.py]:
```suggestion
from .terminal_text_window import TerminalTextWindow

```

File: pilz_teleoperation/src/pilz_teleoperation/curses_key_input.py [hunk 7]
```diff
@@ -44,3 +44,5 @@ def _read_keyboard_input(self):
         key_code = self._screen.getch()
         curses.flushinp()
         return key_code
+
+
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/curses_key_input.py]: to many blanks?

File: pilz_teleoperation/src/pilz_teleoperation/teleoperation_driver.py [hunk 8]
```diff
@@ -88,6 +96,28 @@ def scale_angular_velocity(self, ang_vel_scale):
...
+        else:
+            self.angular.y *= ang_vel_scale
+
+        if self.angular.z == 'max':
+            self.angular.z = 1
+        elif self.angular.z == '-max':
+            self.angular.z = -1
+        else:
+            self.angular.z *= ang_vel_scale
+
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/teleoperation_driver.py]: this is quite redundant.
How about something like:
```
self.angular.x = scale_angular_value(self.angular.x)
self.angular.y = scale_angular_value(self.angular.y)
self.angular.z = scale_angular_value(self.angular.z)

def scale_angular_value(value):
    return 1 if value=='max' else -1 if value=='-max' else value*ang_vel_scale
```

File: pilz_teleoperation/src/pilz_teleoperation/teleoperation_input.py [hunk 9]
```diff
@@ -1,18 +1,3 @@
...
-#
-# This program is free software: you can redistribute it and/or modify
-# it under the terms of the GNU Lesser General Public License as published by
-# the Free Software Foundation, either version 3 of the License, or
-# (at your option) any later version.
-#
-# This program is distributed in the hope that it will be useful,
-# but WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-# GNU Lesser General Public License for more details.
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/teleoperation_input.py]: is this lizence tag removed by intention?

File: pilz_teleoperation/src/pilz_teleoperation/teleoperation_settings.py [hunk 10]
```diff
@@ -19,6 +19,7 @@
 
 FRAMES = ["world", "prbt_tcp"]
 JOINTS = ["prbt_joint_1", "prbt_joint_2", "prbt_joint_3", "prbt_joint_4", "prbt_joint_5", "prbt_joint_6"]
+VELOCITIES = [.2, .6, 1]
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/teleoperation_settings.py]: This is not necessary if the toggle velocity feature is not implemented is it?

File: pilz_teleoperation/src/pilz_teleoperation/teleoperation_settings.py [hunk 11]
```diff
@@ -95,4 +108,6 @@ def __init__(self):
         self.frame_index = 0
         self.joint = JOINTS[0]
         self.joint_index = 0
+        self.velocity_index = 0
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/teleoperation_settings.py]: same goes for the velocity index

File: pilz_teleoperation/src/pilz_teleoperation/terminal_text_window.py [hunk 12]
```diff
@@ -30,14 +30,16 @@ def __init__(self, stdscr, *args, **kwargs):
...
+        self._write_line(4, "Settings:")
+        self._write_line(6, "linear velocity: %.2f m/s \t angular velocity: %.2f rad/s"
+                         % (self._infos.linear_velocity, self._infos.angular_velocity))
+        self._write_line(7, "moves_on_plane:  %s" % self._infos.movement_projection_plane)
+        self._write_line(8, "current_joint:   %s" % self._infos.joint)
+        #self._write_line(7, "rotation_axis:   %s" % (["x", "y", "z"][self._infos.rotation_axis]))
+        self._write_line(9, "target frame:    %s" % self._infos.frame)
+        self._write_line(10, "_______________________________________________________")
+        self._write_line(12, "Controller (keyboard):")
+        self._write_line(13, "fuck + " + self._input_configuration_text)
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/terminal_text_window.py]: "fuck?"

File: pilz_teleoperation/src/pilz_teleoperation/terminal_text_window.py [hunk 13]
```diff
@@ -30,14 +30,16 @@ def __init__(self, stdscr, *args, **kwargs):
...
-        self._write_line(6, "  - current_joint:   %s" % self._infos["joint"])
-        self._write_line(7, "  - target frame:    %s" % (self._infos["target_frame"]))
-        self._write_line(9, "input configuration:")
-        self._write_line(10, self._infos["input_configuration"])
+        self._write_line(4, "Settings:")
+        self._write_line(6, "linear velocity: %.2f m/s \t angular velocity: %.2f rad/s"
+                         % (self._infos.linear_velocity, self._infos.angular_velocity))
+        self._write_line(7, "moves_on_plane:  %s" % self._infos.movement_projection_plane)
+        self._write_line(8, "current_joint:   %s" % self._infos.joint)
+        #self._write_line(7, "rotation_axis:   %s" % (["x", "y", "z"][self._infos.rotation_axis]))
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/terminal_text_window.py]: why is this commented out?

File: pilz_teleoperation/srv/SetTeleopSettings.srv [hunk 14]
```diff
@@ -10,6 +10,10 @@ string TOGGLE_CONTROLLER = TOGGLE_CONTROLLER
 string TOGGLE_TARGET_FRAME = TOGGLE_TARGET_FRAME
 string TOGGLE_JOINT_DOWN = TOGGLE_JOINT_DOWN
 string TOGGLE_JOINT_UP = TOGGLE_JOINT_UP
+string TOGGLE_VELOCITY_UP = TOGGLE_VELOCITY_UP
+string TOGGLE_VELOCITY_DOWN = TOGGLE_VELOCITY_DOWN
```
Comment (DamKoVosh) [inline pilz_teleoperation/srv/SetTeleopSettings.srv]: feature missing or useless as well

File: pilz_teleoperation/test/acceptance_tests/demo_application.launch [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+You should have received a copy of the GNU Lesser General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+-->
+
+<launch>
+    <include file="$(find pilz_teleoperation)/launch/key_teleop.launch" />
+    <node name="test_markers" pkg="pilz_teleoperation" type="demo_application_markers.py" output="screen" />
+</launch>
```
Comment (DamKoVosh) [inline pilz_teleoperation/test/acceptance_tests/demo_application.launch]:
```suggestion
</launch>

```

File: pilz_teleoperation/test/acceptance_tests/demo_application_markers.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+    def set_generic_arrow_marker(self):
+        self.arrow = Marker()
+        self.arrow.action = Marker.ADD
+        self.arrow.header.frame_id = '/world'
+        self.arrow.ns = 'usability_test_arrow'
+        self.arrow.type = Marker.ARROW
+        self.arrow.color.r = 1.0
+        self.arrow.color.g = 1.0
+        self.arrow.color.b = 1.0
+        self.arrow.color.a = 1.0
```
Comment (DamKoVosh) [inline pilz_teleoperation/test/acceptance_tests/demo_application_markers.py]: ...color = Color(1, 1, 1, 1) would be nicer
same in set_custom_text_marker()

File: pilz_teleoperation/test/acceptance_tests/demo_application_markers.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+        self.positions = list()
+        self.positions.append(Point(-.4, -.2, .3))
+        self.positions.append(Point(-.4, .4, .2))
+        self.positions.append(Point(0, .4, .5))
+        self.positions.append(Point(-.4, -.2, .3))
+        self.positions.append(Point(-.4, .2, .3))
+        self.positions.append(Point(-.1, -.3, .1))
+        self.positions.append(Point(-.4, -.2, .3))
+        self.positions.append(Point(-.4, -.2, .3))
+        self.positions.append(Point(-.4, -.2, .3))
```
Comment (DamKoVosh) [inline pilz_teleoperation/test/acceptance_tests/demo_application_markers.py]: self.positions = [
    Point(...)
    ...
]
would mean less redundancy

**Review by DamKoVosh — CHANGES_REQUESTED** [review]

File: pilz_teleoperation/README.md [hunk 18]
```diff
@@ -19,20 +19,21 @@ To use alternative keyboad bindings for the key_teleop driver do:
 Use following Syntax:
     ```
     MovementBindings:
-      'w': <Twist> (e.g. {"linear": {"x": 1.0}})
-      '<CURSES_KEY_NAME>': <Twist> (e.g. KEY_LEFT)
+      'w': <Twist> (e.g. '8': {linear: {x: 1.0, y: -1.0}})
+    JointJogBindings:
+      '<CURSES_KEY_NAME>': <Twist> (e.g. 'KEY_RIGHT': {"velocities": [1]})
```
Comment (DamKoVosh) [inline pilz_teleoperation/README.md]:
```suggestion
      '<CURSES_KEY_NAME>': {"joint_names": ["prbt_joint_2", "prbt_joint_3"], "velocities": [.3, -.1]}
```

File: pilz_teleoperation/README.md (hunk 2 above)
Comment (DamKoVosh) [reply pilz_teleoperation/README.md:31]: point can stay id say, or remove them in the other lines as well

File: pilz_teleoperation/config/keyboard_binding.yaml [hunk 19]
```diff
@@ -60,3 +61,4 @@ Description: |-
   
   numpad : move inside the selected plane
   , : change plane
+
```
Comment (DamKoVosh) [inline pilz_teleoperation/config/keyboard_binding.yaml:64]: why removing it?

File: pilz_teleoperation/src/pilz_teleoperation/teleoperation_driver.py [hunk 20]
```diff
@@ -67,26 +75,17 @@ def scale_linear_velocity(self, vel_scale):
             self.linear.z *= vel_scale
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/teleoperation_driver.py]: please use the new method also for linear velocity

File: pilz_teleoperation/src/pilz_teleoperation/teleoperation_driver.py [hunk 21]
```diff
@@ -213,7 +208,6 @@ def _send_updated_jog(self):
         if self.__key_input_is_new_enough(self.__last_jog_msg):
             new_jog = _TeleoperationJointJog(joint_jog=self.__last_jog_msg)
             new_jog.choose_joint_to_jog(self._settings.joint)
-            new_jog.scale_linear_velocity(self._settings.linear_velocity)
```
Comment (DamKoVosh) [inline pilz_teleoperation/src/pilz_teleoperation/teleoperation_driver.py:216]: this method is renamed above, but is it still used somewhere?
if not, why not?

File: pilz_teleoperation/package.xml [hunk 22]
```diff
@@ -18,11 +18,13 @@
 
   <build_depend>message_generation</build_depend>
   <build_depend>std_msgs</build_depend>
+  <build_depend>visualization_msgs</build_depend>
```
Comment (DamKoVosh) [inline pilz_teleoperation/package.xml]: shouldn't it be run or even test depend? We are not generating any messages with it, but only use it in the demo_app don't we?

**Review by DamKoVosh — CHANGES_REQUESTED** [review]

File: pilz_teleoperation/README.md (hunk 2 above)
Comment (tobi-v) [reply pilz_teleoperation/README.md:31]: Bullet points are all contained in readme, see https://github.com/PilzDE/pilz_teach/blob/usability_test_additions/pilz_teleoperation/README.md

File: pilz_teleoperation/config/keyboard_binding.yaml (hunk 19 above)
Comment (tobi-v) [reply pilz_teleoperation/config/keyboard_binding.yaml:64]: Newline at eof is there, see https://github.com/PilzDE/pilz_teach/blob/usability_test_additions/pilz_teleoperation/config/keyboard_binding.yaml

Comment (tobi-v) [reply pilz_teleoperation/config/keyboard_binding.yaml:64]: Newline at eof is there, see https://github.com/PilzDE/pilz_teach/blob/usability_test_additions/pilz_teleoperation/config/keyboard_binding.yaml

File: pilz_teleoperation/src/pilz_teleoperation/teleoperation_driver.py (hunk 21 above)
Comment (tobi-v) [reply pilz_teleoperation/src/pilz_teleoperation/teleoperation_driver.py:216]: It's used in line 182

**Review by DamKoVosh — APPROVED** [review]
